### PR TITLE
Add batch file (Windows) to convert GOG cp437 config files to UTF-8

### PR DESCRIPTION
**Pull request by Kappa971**
# Description

_Describe a summary of your changes clearly and concisely, including motivation and context._
_Breaking changes need extra explanation on backward compatibility considerations._

_Feel free to include additional details, but please respect the reviewer's time and keep it brief._

As discussed here https://github.com/dosbox-staging/dosbox-staging/pull/3028#issuecomment-1771875459
This is a simple batch file to convert GOG cp437 .conf to UTF-8.

`GOGconf-cp437-to-UTF-8.bat`
```bat
@ECHO OFF

SET conversion_dir=%~dp0
PUSHD "%conversion_dir%"

IF NOT EXIST tools\uconv.exe GOTO readme

ECHO In directory %conversion_dir%:

FOR %%f IN (*.conf) DO (
	CALL :nconv_t "%%f"
)

PAUSE
EXIT /B %ERRORLEVEL%

:nconv_t
ECHO Converting file %~1
tools\uconv -f 437 -t UTF-8 -x nfc "%~1" -o "%~n1.tmp"
MOVE /Y "%~n1.tmp" "%~1" > NUL
EXIT /B 0

:readme
TYPE tools\readme-uconv.txt
ECHO.
PAUSE
EXIT /B
```
`tools\readme-uconv.txt`
```
You need to download the uconv tool (contained in the ICU package):
	https://github.com/unicode-org/icu/releases/latest
Extract uconv.exe (and its dlls) in the "\tools" folder.
```

If requested or necessary, it can of course be improved.
For example I could "generalize" the batch by allowing it to convert a text file with any extension and code page to UTF-8, but I think that's not very useful here... GOG uses DOSBox's .conf files and they are almost certainly cp437 files (right?).
## Related issues

_Related issues can be listed here (remove the section if not applicable.)_

# Manual testing

_Please describe the tests that you ran to verify your changes._

_Provide clear instructions so the reviewers can reproduce and verify your results._

_Include relevant details for your test configuration, operating system, etc._

_Ideally, you would also test your changes with a [sanitizer build](https://github.com/dosbox-staging/dosbox-staging/blob/main/BUILD.md#make-a-sanitizer-build) and confirm no issues were found._

# Checklist

_Please tick the items as you have addressed them. Don't remove items; leave the ones that are not applicable unchecked._

I have:

- [ ] followed the project's [contributing guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/master/CONTRIBUTING.md) and [code of conduct](https://github.com/dosbox-staging/dosbox-staging/blob/master/CODE_OF_CONDUCT.md).
- [ ] performed a self-review of my code.
- [ ] commented on the particularly hard-to-understand areas of my code.
- [ ] split my work into well-defined, bisectable commits, and I [named my commits well](https://github.com/dosbox-staging/dosbox-staging/blob/main/CONTRIBUTING.md#commit-messages).
- [ ] applied the appropriate labels (bug, enhancement, refactoring, documentation, etc.)
- [ ] [checked](https://github.com/dosbox-staging/dosbox-staging/blob/main/scripts/compile_commits.sh) that all my commits can be built.
- [ ] confirmed that my code does not cause performance regressions (e.g., by running the Quake benchmark).
- [ ] added unit tests where applicable to prove the correctness of my code and to avoid future regressions.
- [ ] made corresponding changes to the documentation or the website according to the [documentation guidelines](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md).
- [ ] [locally verified](https://github.com/dosbox-staging/dosbox-staging/blob/main/DOCUMENTATION.md#previewing-documentation-changes-locally) my website or documentation changes.

### CODE REVIEW

File: contrib/scripts/GOGconf-cp437-to-UTF-8.bat [hunk 1]
```diff
@@ -0,0 +1,27 @@
+@ECHO OFF
```
Comment (kcgen) [inline contrib/scripts/GOGconf-cp437-to-UTF-8.bat:1]: Add an SPDX and copyright comment at the top clarify the file's license.

Use `scripts/count-pvs-bugs.py` as an example.

Suggest a bit more explanation about the tool, and can mentioning that the primary use of this is to convert GOG conf files, as an example use case.

Also suggest clarifying that this is .BAT file is meant to be run on the Windoes host, and not insideDOSBox.

File: contrib/scripts/GOGconf-cp437-to-UTF-8.bat [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+ECHO Converting file %~1
+tools\uconv -f 437 -t UTF-8 -x nfc "%~1" -o "%~n1.tmp"
+MOVE /Y "%~n1.tmp" "%~1" > NUL
+EXIT /B 0
+
+:readme
+TYPE tools\readme-uconv.txt
+ECHO.
+PAUSE
+EXIT /B
```
Comment (kcgen) [inline contrib/scripts/GOGconf-cp437-to-UTF-8.bat:27]: Add a carriage return at the end.
(Notice GitHub has flagged this line the 🚫 symbol)

**Review by kcgen — APPROVED** [review]
Nice addition for Windows GOG users, @Kappa971 👍

**Comment by Kappa971** [comment]
> Nice addition for Windows GOG users, @Kappa971 👍

I used the `ComSpec` MS-DOS/Windows environment variable (I noticed this variable in common among all OS) to check if the user is running the batch file on Windows (NT).
This variable on MS-DOS has the value `C:\COMMAND.COM` (`Z:\COMMAND.COM` on DOSBox), while on Windows 9x it is `C:\WINDOWS\COMMAND.COM`. On Windows NT it is `C:\WINDOWS\system32\cmd.exe`.
It's not elegant but it seems to work, and prevents anyone from running this batch on DOSBox. 
Sure there are more sophisticated ways of doing this but I don't know them and doing a web search, they seem unnecessarily complex. If you don't like this solution, I can remove it.

**Comment by OpenRift412** [comment]
Hey, could we perhaps have it search all existing subdirectories for .conf files as well? Would be a lot more convenient for getting a whole bunch of games at once. Would that be doable?

**Comment by Kappa971** [comment]
> Hey, could we perhaps have it search all existing subdirectories for .conf files as well? Would be a lot more convenient for getting a whole bunch of games at once. Would that be doable?

Try this: [conf437-utf8.zip](https://github.com/dosbox-staging/dosbox-staging/files/13060563/conf437-utf8.zip)

I could update the batch file with this:
```bat
@ECHO OFF

REM SPDX-License-Identifier: GPL-2.0-or-later
REM
REM Copyright (C) 2023-2023  The DOSBox Staging Team

REM This batch file converts GOG config files,
REM which use the old 437 code page, to UTF-8.
REM Run this file and enter the root directory of your
REM GOG games (e.g. D:\GAMES).
REM You can also pass the GOG games root directory as a
REM command line argument: 
REM         GOGconf-cp437-to-UTF8 PATH
REM         GOGconf-cp437-to-UTF8 "PATH_WITH_SPACES"

IF NOT "%ComSpec%"=="C:\WINDOWS\system32\cmd.exe" GOTO os_error

PUSHD %~dp0

IF NOT EXIST tools\uconv.exe GOTO readme

IF "%~1"=="" (
    SET /P conversion_dir="Enter the root directory of your GOG games (e.g. D:\GAMES): "
) ELSE (
    SET conversion_dir=%~1
)

ECHO In directory %conversion_dir%:

FOR /R "%conversion_dir%" %%f IN (dosbox*.conf) DO (
	CALL :nconv_t "%%f"
)

PAUSE
EXIT /B %ERRORLEVEL%

:nconv_t
ECHO Converting file %~1
tools\uconv -f 437 -t UTF-8 -x nfc "%~1" -o "%~n1.tmp"
MOVE /Y "%~n1.tmp" "%~1" > NUL
EXIT /B 0

:readme
TYPE tools\readme-uconv.txt
ECHO.
PAUSE
EXIT /B

:os_error
ECHO You are using this batch program on MS-DOS or Windows 9x.
ECHO Please use a more modern version of Windows.
ECHO.
PAUSE
```
@kcgen what do you think?

**Comment by kcgen** [comment]
Maybe the original script could be upgraded to work on directories:
- `/D directory`, non-recursive
- `/R base_directory`, recurse into all sub-directories

**Comment by Kappa971** [comment]
> Maybe the original script could be upgraded to work on directories:
> 
> * `/D directory`, non-recursive
> * `/R base_directory`, recurse into all sub-directories

This batch doesn't check if the source file is already in UTF-8 (I don't know how to do that), so it screws up .conf files already in UTF-8...
If I could find a way to distinguish between UTF-8 and non-UTF-8 files, I don't think anything else would be needed.
You could run the batch file with FOR /R safely and it would do nothing if it doesn't find a not-UTF-8 `dosbox*.conf` file (assuming all non-UTF-8 GOG conf files are cp437).

**Comment by kcgen** [comment]
By "This batch", do you mean the directory-handing one you pasted above or the conf-file based on commited to the repo?

**Comment by Kappa971** [comment]
> By "This batch", do you mean the directory-handing one you pasted above or the conf-file based one commited to the repo?

The new one that I pasted just above.

**Comment by kcgen** [comment]
If the recursive one pasted above will damage any conf that's already utf8, then the user will have some percent of their game confs manged by it. 

So yeah.. that sounds bad.

If you wanted to add a recursive feature, add it to the committed script, and ensure that it skips existing utf8 confs.

**Comment by Kappa971** [comment]
> If the recursive one pasted above will damage any conf that's already utf8, then the user will have some percent of their game confs mangled by it.
> 
> So yeah.. that sounds bad.
> 
> If you wanted to add a recursive feature, add it to the committed script, and ensure that it skips existing utf8 confs.

So, I couldn't find a way to check whether the text file is UTF-8 or not, Windows cmd doesn't have any command for this and I don't know how to do it.
On the other hand, Linux has the `File` tool. I downloaded the Windows version and it detects UTF-8 files.
It's not great to have another program but, unless I can get 3 degrees in computer engineering, I haven't found any other solution 😆

`GOGconf-cp437-to-UTF8.bat`
```bat
@ECHO OFF

REM SPDX-License-Identifier: GPL-2.0-or-later
REM
REM Copyright (C) 2023-2023  The DOSBox Staging Team

REM This batch file converts GOG config files,
REM which use the old 437 code page, to UTF-8.
REM Run this file and enter the root directory of your
REM GOG games (e.g. D:\GAMES).
REM You can also pass the GOG games root directory as a
REM command line argument: 
REM         GOGconf-cp437-to-UTF8 PATH
REM         GOGconf-cp437-to-UTF8 "PATH_WITH_SPACES"

IF NOT "%ComSpec%"=="C:\WINDOWS\system32\cmd.exe" GOTO os_error

setlocal enabledelayedexpansion
PUSHD %~dp0

IF NOT EXIST tools\uconv.exe GOTO readme
IF NOT EXIST tools\icudt*.dll GOTO readme
IF NOT EXIST tools\icuin*.dll GOTO readme
IF NOT EXIST tools\icuuc*.dll GOTO readme
IF NOT EXIST tools\file.exe GOTO readme
IF NOT EXIST tools\magic1.dll GOTO readme
IF NOT EXIST tools\regex2.dll GOTO readme
IF NOT EXIST tools\zlib1.dll GOTO readme
IF NOT EXIST tools\magic GOTO readme

IF "%~1"=="" (
    SET /P conversion_dir="Enter the root directory of your GOG games (e.g. D:\GAMES): "
) ELSE (
    SET conversion_dir=%~1
)

IF NOT EXIST "%conversion_dir%" GOTO dir_not_found

ECHO In directory %conversion_dir%:

FOR /R "%conversion_dir%" %%f IN (dosbox*.conf) DO (
    FOR /F "tokens=*" %%a IN ('tools\file --mime-encoding --brief "%%f" -m tools\magic') DO (
        SET enc=%%a
    )
	IF NOT "!enc!"=="utf-8" CALL :nconv_t "%%f"
)

PAUSE
EXIT /B %ERRORLEVEL%

:nconv_t
ECHO Converting file %~1
tools\uconv -f 437 -t UTF-8 -x nfc "%~1" -o "%~n1.tmp"
MOVE /Y "%~n1.tmp" "%~1" > NUL
EXIT /B 0

:readme
TYPE tools\readme-uconv-file.txt
ECHO.
PAUSE
EXIT /B

:dir_not_found
ECHO Directory not found.
ECHO.
PAUSE
EXIT /B

:os_error
ECHO You are using this batch program on MS-DOS or Windows 9x.
ECHO Please use a more modern version of Windows.
ECHO.
PAUSE
```

`readme-uconv-file.txt`
```
You need to download the following tools:
  - uconv (contained in the ICU package): https://github.com/unicode-org/icu/releases/latest
  - File (Unix tool) for Windows: http://downloads.sourceforge.net/gnuwin32/file-5.03-bin.zip
    and its dependencies: http://downloads.sourceforge.net/gnuwin32/file-5.03-dep.zip
Extract the following files in the "\tools" folder:
    From the ICU zip package                (bin folder): uconv.exe and dlls
    From the File tool zip package          (bin folder): file.exe, magic1.dll
                                     (share\misc folder): magic
    From the File tool dependencies package (bin folder): regex2.dll, zlib1.dll
```

**Comment by kcgen** [comment]
Looks a lot safer!
Feel free to PR the improvement 👍

**Comment by Kappa971** [comment]
Another (final) alternative (besides the one above). 
`GOGconf-cp437-to-UTF8.bat`
```bat
@ECHO OFF

REM SPDX-License-Identifier: GPL-2.0-or-later
REM
REM Copyright (C) 2023-2023  The DOSBox Staging Team

REM This batch file converts GOG config files,
REM which use the old 437 code page, to UTF-8.
REM Syntax:
REM         GOGconf-cp437-to-UTF8 [args] PATH
REM         GOGconf-cp437-to-UTF8 [args] "PATH_WITH_SPACES"
REM [args] can be:
REM -    /R or /r Recursive search through all subdirectories.
REM -    /?       Show the help screen.

IF NOT "%ComSpec%"=="C:\WINDOWS\system32\cmd.exe" GOTO os_error

setlocal enabledelayedexpansion
PUSHD %~dp0

IF NOT EXIST tools\uconv.exe GOTO readme
IF NOT EXIST tools\icudt*.dll GOTO readme
IF NOT EXIST tools\icuin*.dll GOTO readme
IF NOT EXIST tools\icuuc*.dll GOTO readme
IF NOT EXIST tools\file.exe GOTO readme
IF NOT EXIST tools\magic1.dll GOTO readme
IF NOT EXIST tools\regex2.dll GOTO readme
IF NOT EXIST tools\zlib1.dll GOTO readme
IF NOT EXIST tools\magic GOTO readme

SET argCount=0
FOR %%r IN (%*) DO (
    SET /A argCount+=1
)
IF %argCount% GTR 2 (
    GOTO help
)

IF NOT "%~2"=="" IF /I NOT "%~1"=="/R" (
    GOTO help
)

IF "%~1"=="/?" (
    GOTO help
)

IF /I "%~1"=="/R" (
    SET recursive=true
) ELSE (
    SET recursive=false
)

IF "%~1"=="" (
    CALL :menu
) ELSE (
    SET conversion_dir=%~1
)

IF NOT "%~2"=="" (
    SET conversion_dir=%~2
)

IF NOT EXIST "%conversion_dir%" (
    GOTO dir_not_found
)

ECHO In directory %conversion_dir%:

IF "%recursive%"=="false" (
    CALL :normal_search
) ELSE (
    CALL :recursive_search
)

PAUSE
EXIT /B %ERRORLEVEL%

:menu
ECHO 1. Normal search from the directory of your GOG game (e.g. D:\GAMES\DOOM)
ECHO 2. Recursive search from the root directory of your GOG games (e.g. D:\GAMES)
CHOICE /C 12 /N /M "Enter choice (1,2): "
IF "%ERRORLEVEL%"=="1" SET recursive=false
IF "%ERRORLEVEL%"=="2" SET recursive=true
SET /P conversion_dir="Enter path: "
ECHO.
EXIT /B 0

:normal_search
SET count=0
SET countSkipped=0
FOR %%f IN ("%conversion_dir%\dosbox*.conf") DO (   
    FOR /F "tokens=*" %%a IN ('tools\file --mime-encoding --brief "%%f" -m tools\magic') DO (
        SET enc=%%a
    )
	IF NOT "!enc!"=="utf-8" CALL :nconv_t "%%f"
    IF NOT "!enc!"=="utf-8" SET /A count+=1
    IF "!enc!"=="utf-8" SET /A countSkipped+=1
)
IF %count% GTR 0 (
    ECHO %count% file/s converted, %countSkipped% file/s skipped.
) ELSE (
    ECHO No files converted.
)
EXIT /B 0

:recursive_search
SET count=0
SET countSkipped=0
FOR /R "%conversion_dir%" %%f IN (dosbox*.conf) DO (
    FOR /F "tokens=*" %%a IN ('tools\file --mime-encoding --brief "%%f" -m tools\magic') DO (
        SET enc=%%a
    )
	IF NOT "!enc!"=="utf-8" CALL :nconv_t "%%f"
    IF NOT "!enc!"=="utf-8" SET /A count+=1
    IF "!enc!"=="utf-8" SET /A countSkipped+=1
)
IF %count% GTR 0 (
    ECHO %count% file/s converted, %countSkipped% file/s skipped.
) ELSE (
    ECHO No files converted.
)
EXIT /B 0

:nconv_t
ECHO Converting file %~1
tools\uconv -f 437 -t UTF-8 -x nfc "%~1" -o "%~n1.tmp"
MOVE /Y "%~n1.tmp" "%~1" > NUL
EXIT /B 0

:readme
TYPE tools\readme-gog437.txt
ECHO.
PAUSE
EXIT /B

:dir_not_found
ECHO Directory not found.
ECHO.
PAUSE
EXIT /B

:help
ECHO This batch file converts GOG config files,
ECHO which use the old 437 code page, to UTF-8.
ECHO Syntax:
ECHO         GOGconf-cp437-to-UTF8 [args] PATH
ECHO         GOGconf-cp437-to-UTF8 [args] "PATH_WITH_SPACES"
ECHO [args] can be:
ECHO -    /R or /r Recursive search through all subdirectories.
ECHO -    /?       Show the help screen.
EXIT /B

:os_error
ECHO You are using this batch program on MS-DOS or Windows 9x.
ECHO Please use a more modern version of Windows.
ECHO.
PAUSE
```
I don't know if you like it, maybe for you this batch may be "rude" 😁, but it seems to work (I did several tests, creating a directory tree with some cp437 dosbox*.conf files). However I would be more confident if someone else tried it.
[conf437-utf8.zip](https://github.com/dosbox-staging/dosbox-staging/files/13063713/conf437-utf8.zip)

@kcgen, the choice is yours, whether this or the one above (or neither, and it remains the current one in the repository).

**Pull request merged**